File: code/ex16.py
class Solution:
    def threeSumClosest(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """

        length = len(nums)
        minimum = 2 **31

        for i in range(0, length - 2):
            for j in range(i + 1, length - 1):
                for k in range(j + 1, length):
                    print(nums[i], nums[j], nums[k])
                    gap = abs(nums[i] + nums[j] + nums[k] - target)

                    if gap < minimum:
                        minimum = abs(nums[i] + nums[j] + nums[k] - target)
                        r = nums[i] + nums[j] + nums[k]


        return r

sol = Solution()
print(sol.threeSumClosest([1, 1, -1, -1, 3],
                          3))

File: code/ex16.cpp
class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        sort(nums.begin(), nums.end());
        
        int n = nums.size();
        int res = 0, diff = INT_MAX;
        
        for (int i = 0; i < n - 2; i++) {
            int remain = target - nums[i];
            
            // Find two num whose sum is closest to ramain
            int lo = i + 1, hi = n - 1;
            while (lo < hi) {
                int sum = nums[i] + nums[lo] + nums[hi];
                if (abs(sum - target) < diff) {
                    diff = abs(sum - target);
                    res = sum;
                }
                if (nums[lo] + nums[hi] < remain)    lo++;
                else    hi--;
            }
        }
        
        return res;
    }
};